File: Data Structures & Algorithms/validate-parentheses/submission-1.java
class Solution {
    public boolean isValid(String s) {
        // 使用 ArrayDeque 作為 Stack 
        Deque<Character> stack = new ArrayDeque<>();
        
        for (char c : s.toCharArray()) {
            if (c == '(') stack.push(')');
            else if (c == '{') stack.push('}');
            else if (c == '[') stack.push(']');
            // 如果是右括號，直接跟彈出的元素比對
            else if (stack.isEmpty() || stack.pop() != c) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}

File: Data Structures & Algorithms/interleaving-string/submission-1.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        i1 = 0
        i2 = 0

        if len(s1) + len(s2) != len(s3):
            return False
        
        for c in s3:
            c1 = "" if i1 == len(s1) else s1[i1]
            c2 = "" if i2 == len(s2) else s2[i2]

            if c == c2:
                i2 += 1
            elif c == c1:
                i1 += 1
            else:
                return False
        
        return True
